File: src/main/java/de/huddeldaddel/euler/math/TriangularNumberSequence.java
package de.huddeldaddel.euler.math;

import java.math.BigInteger;
import java.util.function.Supplier;

public class TriangularNumberSequence implements Supplier<BigInteger> {

    private int index = 0;
    private BigInteger last = BigInteger.ZERO;

    public BigInteger nextValue() {
        index++;
        return last = last.add(BigInteger.valueOf(index));
    }

    @Override
    public BigInteger get() {
        return nextValue();
    }

}

File: src/main/kotlin/de/huddeldaddel/euler/math/TriangularNumberSequence.kt
package de.huddeldaddel.euler.math

import java.util.function.Supplier

class TriangularNumberSequence : Supplier<Long> {

    private var index = 0
    private var last = 0L

    fun nextValue(): Long {
        index++
        last += index
        return last
    }

    override fun get(): Long {
        return nextValue()
    }

}
